---
title: "Clean RosMap Meta Data"
output: html_document
---

```{r setup, include = FALSE}
options(xtable.type="html")
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

1. [Write your own config file](https://kelshmo.github.io/sageseqr/articles/customize-config.html).

2. Specify the active configuration by setting `R_CONFIG_ACTIVE`.

3. Load the `sageseqr` library and login to [Synapse](https://www.synapse.org/). `rnaseq_plan()` calls the arguments from the config file and creates the `drake` plan. Execute `drake::make(plan)` to compute. Run this code from your project root.

```{r run-plan}
#library(devtools)
#devtools::install_git('https://github.com/kelshmo/sageseqr.git')
#devtools::install_github('th1vairam/CovariateAnalysis@dev')
#remotes::install_github("Sage-Bionetworks/sageseqr")
library(sageseqr)
library(edgeR)
library(ggplot2)
library(CovariateAnalysis) #get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(plyr)
library(tidyverse)
library(psych)
library(limma)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)
library(glmnet)
library(knitr)
library(doParallel)
library(foreach)
library(githubr)
#BiocManager::install("WGCNA")
# Login to Synapse. Make a Synapse account and use synapser to login: https://r-docs.synapse.org/articles/manageSynapseCredentials.html

synapser::synLogin()

#Seq Data Here:
#https://www.synapse.org/#!Synapse:syn21788344/files/
```
```{r PullAndMergeV2}
##RNA-Seq Covariates:
library(synapser)
RNASeq <- read.csv(synGet('syn21088596')$path)
RNASeq <- RNASeq[ !duplicated(RNASeq),]
RNASeq[ RNASeq$specimenID == '02_120405',]

trans <- c(TRUE,FALSE)
names(trans) <- c('True','False')

DC1 <- RNASeq[ as.character(RNASeq$notes) %in% "data contribution batch 1" &  (as.character(RNASeq$sequencingBatch) %in% 'nan') ==F, ] 
DC1$libraryBatch <- paste0( '1_', as.character( DC1$libraryBatch ) ) 
DC1$sequencingBatch <- paste0( '1_', as.character( DC1$sequencingBatch ) ) 
DC1$notes <- 'DataCut_1'

DC1$isStranded <- trans[as.character(DC1$isStranded)]  

DC2 <- RNASeq[ (as.character(RNASeq$notes) %in% "data contribution batch 2")==T, ]
DC2$sequencingBatch <- paste0( '2_', gsub('NYGC', '', as.character(DC2$sequencingBatch)))
DC2$notes <- 'DataCut_2'
DC2$isStranded <- trans[as.character(DC2$isStranded)]  

DC3 <- RNASeq[ (as.character(RNASeq$notes) %in% "data contribution batch 3")==T, ]
DC3$sequencingBatch <- paste0( '3_', gsub('RISK_', '', as.character(DC3$sequencingBatch)))
DC3$notes <- 'DataCut_3'


rnaseq <- as.data.frame(rbind(DC1,DC2,DC3))

#Non-Variant:
##isStranded, runType
#Never used or redundant:
##rnaBatch, libraryBatch

#Clean out useless Data:
rnaseq <- rnaseq[ , (colnames(rnaseq) %in% c('isStranded', 'runType', 'rnaBatch', 'libraryBatch' ))==F]
rnaseq$platform <- as.factor(as.character(rnaseq$platform))
rnaseq$sequencingBatch <- as.factor(as.character(rnaseq$sequencingBatch))
rnaseq$libraryPrep <- as.factor(as.character(rnaseq$libraryPrep))
rnaseq$libraryPreparationMethod <- as.factor(as.character(rnaseq$libraryPreparationMethod))
rnaseq$readStrandOrigin <- as.factor(as.character(rnaseq$readStrandOrigin))
rnaseq$readLength <- as.factor(as.character(rnaseq$readLength))
rnaseq$notes <- as.factor(as.character(rnaseq$notes))
rnaseq$specimenID <- as.character(rnaseq$specimenID)

table(rnaseq$specimenID)[ table(rnaseq$specimenID) > 1 ] 
rnaseq[ (rnaseq$specimenID %in% '492_120515') , ]

##Biospecimin Covariates:
Bio <- read.csv(synGet('syn21323366')$path)
Bio$specimenID <- as.character(Bio$specimenID)
Bio <- Bio[ Bio$specimenID %in% rnaseq$specimenID, ]

table(Bio$samplingAge)
#Non-Variant:
##organ, cellType, fastingState, isPostMortem
#Never used or redundant:
##specimenIdSource, samplingDate, BrodmannArea, sampleStatus, Bio$tissueWeight, tissueVolume, nucleicAcidSource, Bio$samplingAge

Bio <- Bio[ , colnames(Bio)[(colnames(Bio) %in% c('specimenIdSource', 'samplingDate', 'BrodmannArea', 'sampleStatus', 'tissueWeight', 'tissueVolume', 'nucleicAcidSource', 'samplingAge', 'organ', 'cellType', 'fastingState', 'isPostMortem'))==F] ]

Bio$tissue<-as.character(Bio$tissue)
trans <- c( 'DLPFC', 'ACC', 'PCC' )
names(trans) <- names( table(Bio$tissue) )
Bio$tissue <- trans[Bio$tissue]
Bio$tissue <- as.factor(Bio$tissue)

row.names(Bio) <- Bio$specimenID

Total_Meta <- cbind( rnaseq[ ,1:dim(rnaseq)[2] ], Bio[ rnaseq$specimenID ,1:dim(Bio)[2] ])

#Clinical Meta Data:
##syn3191087
clinical <- read.csv(synGet('syn3191087')$path)

row.names(clinical) <- as.character(clinical$individualID)
Total_Meta <- cbind( Total_Meta[ ,1:dim(Total_Meta)[2] ], clinical[ as.character(Total_Meta$individualID) ,1:dim(clinical)[2] ])


#SeqStats
ss1 <- read.table(synGet('syn22283384')$path, header =T, sep='\t')
ss2 <- read.table(synGet('syn22301603')$path, header =T, sep='\t')
ss3 <- read.table(synGet('syn22314232')$path, header =T, sep='\t')

SS <- rbind(ss1,ss2,ss3)
row.names(SS) <- SS$sample

Total_Meta <- cbind( Total_Meta[ ,1:dim(Total_Meta)[2] ], SS[ as.character(Total_Meta$specimenID) ,1:dim(SS)[2] ])
#_#Total_Meta <- Total_Meta[ , colnames(Total_Meta)[ (colnames(Total_Meta) %in% c('specimenID.1','individualID.1', 'sample')) ==F]]
table(Total_Meta$tissue)

Total_Meta <- Total_Meta[ !duplicated(Total_Meta$specimenID), ]

#Get Count Data
Ros_DC1 <- get_data( 'syn22283382' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t()

Ros_DC2 <- get_data( 'syn22301601' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t() 

Ros_DC3 <- get_data( 'syn22314230' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t()

total_counts <- cbind( Ros_DC1, Ros_DC2, Ros_DC3 )

table( colnames(total_counts) %in% Total_Meta$specimenID )

total_counts <- total_counts[,Total_Meta$specimenID]

####
#Remove Identity Issues:
Toss <- read.csv(synGet('syn22355545')$path)


#Total_Meta[ Total_Meta$projid %in% 24680888, ]
#Toss[ Toss$Projid %in% 24680888, ]

#Toss$ExternalID[ Toss$ExternalID %in% Total_Meta$specimenID ]

#Total_Meta[ Total_Meta$projid %in% Toss[ ( Toss$ExternalID %in% Total_Meta$specimenID )==F, ]$Projid , 1:10] 

#'Sample_R12365619-AC' %in% Total_Meta$specimenID
#'Sample_R21162197-DLPFC' %in% Total_Meta$specimenID
#table( grepl( '21162197', Total_Meta$specimenID ) )

#toss <- Toss[ , ]

table(rnaseq$specimenID %in% Toss$ExternalID)

# Fix Broken Samples
Toss$SageID <- NA
Toss[ Toss$ExternalID %in% Total_Meta$specimenID, ]$SageID <- Toss[ Toss$ExternalID %in% Total_Meta$specimenID, ]$ExternalID
Toss$tissue <- NA

for( i in 1:dim(Toss)[1] ){
  #Pull Tissue based on anotation in meta data or the sample name if its diff btwn the Swap and meta data sample IDs
  if( is.na(Toss[i,]$SageID) ){
    if( isTRUE(grepl('-DLPFC', Toss[i,]$ExternalID )) ){
      Toss[ i, ]$tissue <- 'DLPFC'
    }else{
      if( isTRUE(grepl('-PCC', Toss[i,]$ExternalID )) ){
        Toss[ i, ]$tissue <- 'PCC'
      }else{
        Toss[ i, ]$tissue <- 'ACC'
      }
    }
  #Pull swapped Sample IDs based on Tissue and Project Indv ID
  Toss[i,]$SageID <- Total_Meta[ Total_Meta$projid %in% Toss[i,]$Projid & as.character(Total_Meta$tissue) %in% Toss[i,]$tissue, ]$sample[1]
  }else{
    Toss[ i, ]$tissue <- as.character(Total_Meta[ Total_Meta$specimenID %in% Toss[i,]$SageID, ]$tissue)
  }
}

###Add the second sample of the 
Toss[ 1+dim(Toss)[1], ] <- Toss[ Toss$SageID %in% '433_120507' , ]
Toss[ dim(Toss)[1], ]$SageID <- 'Sample_R4361022-DLPFC'

#_# METADATA[ METADATA$sampleID %in% c('Sample_R4361022-DLPFC', '433_120507'), ]


# "Sample_R6796261-DLPFC" "RISK_240_rerun"
#Sample_MAP65838800-DLPFC 65838800   56432243 Sample_R6796261-DLPFC  DLPFC


table(Toss$SageID %in% Total_Meta$sample)
table(Toss$Projid %in% Total_Meta$projid)

```


```{r PullAndMerge }
metadata <- Total_Meta
colnames(metadata)[ colnames(metadata) == 'specimenID' ] <- 'Sampleid'
colnames(metadata)[ colnames(metadata) == 'RIN' ] <- 'RINcontinuous'

counts <- total_counts
# Identify variables with missing data
removed_samples <- metadata$Sampleid[is.na(metadata$cogdx) |
                                 is.na(metadata$braaksc) |
                                 is.na(metadata$ceradsc) |
                                 is.na(metadata$RINcontinuous) |
                                 is.na(metadata$pmi) |
                                 is.na(metadata$RnaSeqMetrics__INTRONIC_BASES) |
                                 is.na(metadata$age_death)]

# Pick higher quality RIN batch
metadata <- metadata %>%
  dplyr::group_by(Sampleid) %>%
  dplyr::top_n(1, RINcontinuous)

# Remove variables with missing data
metadata <- metadata %>%
  ungroup %>%
  dplyr::filter(Sampleid %in% colnames(counts)) %>%
  dplyr::filter(!(Sampleid %in% removed_samples))

# Remove samples that do not meet criteria
counts <- counts[,metadata$Sampleid]


# Harmonize case-control status
metadata$diagnosis <- "other"
metadata$diagnosis[metadata$cogdx == 1 & metadata$braaksc <= 3 & metadata$ceradsc >= 3] <- "control"
metadata$diagnosis[metadata$cogdx == 4 & metadata$braaksc >= 4 & metadata$ceradsc <= 2] <- "AD"

# Add sex variable
metadata$sex <- "female"
metadata$sex[metadata$msex == 1] <- "male"

# Add tissue variable
metadata$tissue <- "DLPFC"

# Add APOE4 genotype = 0,1,2
metadata$APOE4 <- 0
metadata$APOE4[metadata$apoe_genotype %in% c(24,34)] <- 1
metadata$APOE4[metadata$apoe_genotype %in% c(44)] <- 2

# Compute square of RIN
metadata$RIN2 <- metadata$RINcontinuous^2
colnames(metadata)[ colnames(metadata) == 'sequencingBatch' ] <- 'Batch'

# Filter for variables to be used in downstream analysis as covariates
covariates <- c("Sampleid","projid","cogdx", "diagnosis", "tissue", "APOE4", "Batch", "sex", "race", "spanish", "cogdx", "APOE4",
                "RINcontinuous", "RIN2", "age_death", "pmi", "educ", "AlignmentSummaryMetrics__PCT_PF_READS_ALIGNED",
                "RnaSeqMetrics__PCT_CODING_BASES","RnaSeqMetrics__PCT_INTERGENIC_BASES", "RnaSeqMetrics__PCT_INTRONIC_BASES",
                "RnaSeqMetrics__PCT_RIBOSOMAL_BASES")

#_# metadata <- dplyr::select(metadata, covariates)

# Clean variable names
colnames(metadata) <- gsub("RnaSeqMetrics__", "", colnames(metadata))
colnames(metadata) <- gsub("AlignmentSummaryMetrics__", "", colnames(metadata))
colnames(metadata) <- tolower(colnames(metadata))

```

### Sex Check
```{r SexCheck }
COUNT <- as.data.frame( counts )
COUNT$ensembl_gene_id <- do.call(rbind, strsplit(row.names(COUNT), '\\.'))[,1]
COUNT <- COUNT[ !duplicated(COUNT$ensembl_gene_id), ]
row.names(COUNT) <- COUNT$ensembl_gene_id
COUNT <- COUNT[ , colnames(COUNT) != 'ensembl_gene_id' ]
#COUNT <- COUNT[ grepl('ENSG', row.names(COUNT)), ]

library(biomaRt)
#ensembl_gene_id  hgnc_symbol percentage_gene_gc_content  gene_biotype  chromosome_name gene_length
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
biomart_results <- getBM(filters= "ensembl_gene_id", attributes= c('ensembl_gene_id', 'hgnc_symbol',  'percentage_gene_gc_content', 'gene_biotype', 'chromosome_name',  "transcript_length"),values=c( "ENSG00000229807", 'ENSG00000183878' ), mart= mart)

foo <- biomart_results[ biomart_results[ biomart_results$hgnc_symbol == 'XIST',]$transcript_length == max( biomart_results[biomart_results$hgnc_symbol == 'XIST',]$transcript_length), ]
biomart <- foo[ foo$chromosome_name == 'X', ]
foo <- biomart_results[ biomart_results[ biomart_results$hgnc_symbol == 'UTY',]$transcript_length == max( biomart_results[ biomart_results$hgnc_symbol == 'UTY',]$transcript_length), ]
biomart <- rbind(biomart, foo[ foo$chromosome_name == 'Y', ])

biomart_results <- biomart
colnames(biomart_results)[ colnames(biomart_results) == 'transcript_length' ] <- 'gene_length'
COUNT <- COUNT[ row.names(COUNT) %in% biomart_results$ensembl_gene_id, ]
row.names(biomart_results) <- biomart_results$ensembl_gene_id

METADATA <- as.data.frame( metadata )
colnames(METADATA)[ colnames(METADATA) == 'sampleid' ] <- 'sampleID'

#Remove JiShu Swaps (N=28)
#_# METADATA <- METADATA[ (METADATA$sampleID %in% Toss$ExternalID) == F, ] 
#_# COUNT <- COUNT[ , ( colnames(COUNT) %in% Toss$ExternalID ) == F ] 

#Remove all samps from the projid
#_# METADATA <- METADATA[ (METADATA$projid %in% Toss$Projid) == F, ] 
#_# COUNT <- COUNT[ , colnames(COUNT) %in% METADATA$sampleID ] 

#Remove all samps from the new.projid
#_# METADATA <- METADATA[ (METADATA$projid %in% Toss$new.projid) == F, ] 
#_# COUNT <- COUNT[ , colnames(COUNT) %in% METADATA$sampleID ] 

#UnCleaned Metadata
row.names(METADATA) <- METADATA$sampleID
COUNT$ensembl_gene_id <- row.names(COUNT)
REPORTED.GENDER.COUNTS = biomart_results %>% 
  left_join(COUNT) %>%
  dplyr::select(-one_of("percentage_gene_gc_content")) %>%
  filter(chromosome_name == "X" |chromosome_name == "Y") %>% 
  tidyr::gather(key = item, value = value, -c( ensembl_gene_id, hgnc_symbol, gene_biotype, chromosome_name, gene_length, ensembl_gene_id)) %>%
  dplyr::mutate(value = log(value)) %>%
  dplyr::rename(`counts(log)`= value) %>% 
  dplyr::rename(sampleID = item) %>%
  left_join(METADATA[,c("sampleID", "sex")]) %>% 
  dplyr::rename(`Reported Gender` = sex) 
my.theme <- theme_bw() %+replace% theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1), plot.title=element_text(hjust=0.5))
p = list()
p[[1]] = ggplot(filter(REPORTED.GENDER.COUNTS, chromosome_name == "X"), aes(x = `Reported Gender`, y = `counts(log)`)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('X') + my.theme
p[[2]] = ggplot(filter(REPORTED.GENDER.COUNTS, chromosome_name == "Y"), aes(x = `Reported Gender`, y = `counts(log)`)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('Y') + my.theme
multiplot(plotlist = p, cols = 2)
##XIST and UTY expression 
#ENSG00000229807.11 and ENSG00000183878.15 
#Plot initial data
FILT <- REPORTED.GENDER.COUNTS[ , c('ensembl_gene_id', 'chromosome_name', 'sampleID','counts(log)', 'Reported Gender')] %>% 
  filter( ensembl_gene_id == "ENSG00000229807" | ensembl_gene_id == "ENSG00000183878") %>% 
  dplyr::select(-one_of("chromosome_name")) %>% 
  tidyr::spread(key = ensembl_gene_id, value = `counts(log)`) %>% 
  mutate(XIST = as.numeric(`ENSG00000229807`)) %>% 
  mutate(UTY = as.numeric(`ENSG00000183878`)) %>% 
  mutate(UTY = ifelse(UTY == -Inf, 0, UTY)) %>% 
  mutate(XIST = ifelse(XIST == -Inf, 0, XIST))
p = ggplot(FILT, aes (x= XIST, y = UTY)) 
p = p + geom_point(aes(color=`Reported Gender`)) + 
  ggtitle("Sex Check Inital Sex: RosMap Combined Cohort") + 
  theme(plot.title = element_text(hjust = 0.5, size = 15)) +
  labs(colour = "Reported Gender")
p


table(FILT[ FILT$UTY > 3.7 ,]$`Reported Gender`)
# female   male 
#     14    868 
# 100*( 14/34.49765 )

table(FILT[ FILT$UTY < 3.7 ,]$`Reported Gender`)
# female   male 
#   1655     11
# 100*( 11/65.50235 )



#Remove the Toss$SageID
row.names(METADATA) <- METADATA$sampleID
COUNT$ensembl_gene_id <- row.names(COUNT)
REPORTED.GENDER.COUNTS = biomart_results %>% 
  left_join(COUNT[ , (colnames(COUNT) %in% Toss$SageID) == F ]) %>%
  dplyr::select(-one_of("percentage_gene_gc_content")) %>%
  filter(chromosome_name == "X" |chromosome_name == "Y") %>% 
  tidyr::gather(key = item, value = value, -c( ensembl_gene_id, hgnc_symbol, gene_biotype, chromosome_name, gene_length, ensembl_gene_id)) %>%
  dplyr::mutate(value = log(value)) %>%
  dplyr::rename(`counts(log)`= value) %>% 
  dplyr::rename(sampleID = item) %>%
  left_join(METADATA[ (row.names(METADATA) %in% Toss$SageID)==F ,c("sampleID", "sex")]) %>% 
  dplyr::rename(`Reported Gender` = sex) 
my.theme <- theme_bw() %+replace% theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1), plot.title=element_text(hjust=0.5))
p = list()
p[[1]] = ggplot(filter(REPORTED.GENDER.COUNTS, chromosome_name == "X"), aes(x = `Reported Gender`, y = `counts(log)`)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('X') + my.theme
p[[2]] = ggplot(filter(REPORTED.GENDER.COUNTS, chromosome_name == "Y"), aes(x = `Reported Gender`, y = `counts(log)`)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('Y') + my.theme
multiplot(plotlist = p, cols = 2)
##XIST and UTY expression 
#ENSG00000229807.11 and ENSG00000183878.15 
#Plot initial data
FILT <- REPORTED.GENDER.COUNTS[ , c('ensembl_gene_id', 'chromosome_name', 'sampleID','counts(log)', 'Reported Gender')] %>% 
  filter( ensembl_gene_id == "ENSG00000229807" | ensembl_gene_id == "ENSG00000183878") %>% 
  dplyr::select(-one_of("chromosome_name")) %>% 
  tidyr::spread(key = ensembl_gene_id, value = `counts(log)`) %>% 
  mutate(XIST = as.numeric(`ENSG00000229807`)) %>% 
  mutate(UTY = as.numeric(`ENSG00000183878`)) %>% 
  mutate(UTY = ifelse(UTY == -Inf, 0, UTY)) %>% 
  mutate(XIST = ifelse(XIST == -Inf, 0, XIST))
p = ggplot(FILT, aes (x= XIST, y = UTY)) 
p = p + geom_point(aes(color=`Reported Gender`)) + 
  ggtitle("Sex Check Inital Sex: RosMap Combined Cohort") + 
  theme(plot.title = element_text(hjust = 0.5, size = 15)) +
  labs(colour = "Reported Gender")
p

table(FILT[ FILT$UTY > 3.7 ,]$`Reported Gender`)
# female   male 
#     0    857 
# 100*( 14/34.49765 )
table(FILT[ FILT$UTY < 3.7 ,]$`Reported Gender`)
# female   male 
#   1623     1
# 100*( 11/65.50235 )

FILT[ FILT$UTY < 4 & FILT$`Reported Gender` == 'male' ,]

#Toss <- Toss[,c('ExternalID', 'Projid', 'new.projid', 'SageID', 'tissue')]
Toss$Projid.Sex <- NA
Toss$new.projid.Sex <- NA
for( i in 1:dim(Toss)[1] ){
  Toss[ i, ]$Projid.Sex <- clinical[ clinical$projid %in% Toss[i,]$Projid,]$msex
  Toss[ i, ]$new.projid.Sex <- clinical[ clinical$projid %in% Toss[i,]$new.projid,]$msex
}

Toss <- as.data.frame( Toss )
#Add the last sex swap
Toss[ dim(Toss)[1]+1, ] <- NA

Toss[ dim(Toss)[1], ]$SageID <- FILT[ FILT$UTY < 4 & FILT$`Reported Gender` == 'male' ,]$sampleID
Toss[ dim(Toss)[1], ]$ExternalID <- FILT[ FILT$UTY < 4 & FILT$`Reported Gender` == 'male' ,]$sampleID
Toss[ dim(Toss)[1], ]$Projid <- METADATA[ METADATA$sampleID == FILT[ FILT$UTY < 4 & FILT$`Reported Gender` == 'male' ,]$sampleID,]$projid
Toss[ dim(Toss)[1], ]$Projid.Sex <- 1
Toss[ dim(Toss)[1], ]$tissue <- as.character( Total_Meta[ Total_Meta$specimenID == METADATA[ METADATA$sampleID == FILT[ FILT$UTY < 4 & FILT$`Reported Gender` == 'male' ,]$sampleID,]$sampleID, ]$tissue )


thisFileName <- '01_RosMap_MetaData_Wrangle.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/RosMap_RNASeq", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))

write.csv( Toss,'SampleSwaps.csv', row.names = F )
activity <- Activity(
    'Sample Swap Issues',
    executed = thisFile,
    description='samples to exclude due to not matching patient',
    used=c('syn22355545',"syn22314230","syn22301601","syn22283382","syn22283384","syn22301603","syn22314232","syn3191087",'syn21323366','syn21088596'))
file <- File('SampleSwaps.csv', description='Sample Swap Info', parent='syn23569327')
file <- synStore(file, activity=activity)

sink_metadata <- metadata 
sink_Total_Meta <- Total_Meta

Total_Meta <- Total_Meta[ , ( colnames(Total_Meta) %in% c('specimenID.1', 'individualID.1') ) == F ]
Meta_Final <- Total_Meta

#Add Cleaned Vals
Meta_Final <- Meta_Final[ Meta_Final$specimenID %in% metadata$sampleid, ]

row.names(Meta_Final) <- Meta_Final$specimenID
row.names(metadata) <- metadata$sampleid
metadata <- metadata[row.names(Meta_Final),]


Missing <- colnames(metadata)[ (colnames(metadata) %in% colnames(Meta_Final)) == F ]
Missing <- Missing[ grepl('.1', Missing)==F]
for( name in Missing ){
  eval( parse( text = paste0( 'Meta_Final$', name, ' <- metadata$', name ) ))
}
Meta_Final$FailQC_SampleSwapped <- 0
Meta_Final[ Meta_Final$specimenID %in% Toss$SageID, ]$FailQC_SampleSwapped <- 1

#Fix IDs with trimmed Zeros
table( sapply(Meta_Final$projid[!duplicated(Meta_Final$projid)], nchar) )
Meta_Final$projid_repaired <- stringr::str_pad(Meta_Final$projid, 8, side = "left", pad = 0)
#table( sapply(FOO[!duplicated(FOO)], nchar) )


ORD <- c( "specimenID", "sampleid", "sample", "individualID", "individualid", "projid", "projid_repaired", "Study", "study", "sequencingBatch", "notes", "batch", "apoe_genotype", "apoe4", "pmi", "rincontinuous", "rin2", "RIN",  "FailQC_SampleSwapped", "readStrandOrigin", "readstrandorigin", "readLength", "readlength", "platform", "libraryPrep", "libraryPreparationMethod","librarypreparationmethod", "libraryprep", "tissue", "diagnosis", "msex", "sex", "educ", "race", "spanish", "age_at_visit_max", "age_first_ad_dx", "age_death", "cts_mmse30_first_ad_dx", "cts_mmse30_lv", "braaksc", "ceradsc", "cogdx", "dcfdx_lv", "AlignmentSummaryMetrics__BAD_CYCLES", "bad_cycles", "AlignmentSummaryMetrics__CATEGORY", "category", "AlignmentSummaryMetrics__MEAN_READ_LENGTH", "mean_read_length", "AlignmentSummaryMetrics__PCT_ADAPTER", "pct_adapter", "AlignmentSummaryMetrics__PCT_CHIMERAS", "pct_chimeras", "AlignmentSummaryMetrics__PCT_PF_READS", "pct_pf_reads", "AlignmentSummaryMetrics__PCT_PF_READS_ALIGNED", "AlignmentSummaryMetrics__PCT_READS_ALIGNED_IN_PAIRS", "pct_reads_aligned_in_pairs", "AlignmentSummaryMetrics__PF_ALIGNED_BASES", "pf_aligned_bases", "AlignmentSummaryMetrics__PF_HQ_ALIGNED_BASES", "pf_hq_aligned_bases", "AlignmentSummaryMetrics__PF_HQ_ALIGNED_Q20_BASES", "pf_hq_aligned_q20_bases", "AlignmentSummaryMetrics__PF_HQ_ALIGNED_READS", "pf_hq_aligned_reads", "AlignmentSummaryMetrics__PF_HQ_ERROR_RATE", "pf_hq_error_rate", "AlignmentSummaryMetrics__PF_HQ_MEDIAN_MISMATCHES", "pf_hq_median_mismatches" , "AlignmentSummaryMetrics__PF_INDEL_RATE",  "pf_indel_rate" , "AlignmentSummaryMetrics__PF_MISMATCH_RATE", "pf_mismatch_rate", "AlignmentSummaryMetrics__PF_NOISE_READS", "pf_noise_reads", "AlignmentSummaryMetrics__PF_READS", "pf_reads", "AlignmentSummaryMetrics__PF_READS_ALIGNED", "pf_reads_aligned", "AlignmentSummaryMetrics__READS_ALIGNED_IN_PAIRS", "reads_aligned_in_pairs", "AlignmentSummaryMetrics__STRAND_BALANCE", "strand_balance", "AlignmentSummaryMetrics__TOTAL_READS", "total_reads", "RnaSeqMetrics__CODING_BASES", "coding_bases", "RnaSeqMetrics__CORRECT_STRAND_READS", "correct_strand_reads", "RnaSeqMetrics__IGNORED_READS", "ignored_reads", "RnaSeqMetrics__INCORRECT_STRAND_READS", "incorrect_strand_reads", "RnaSeqMetrics__INTERGENIC_BASES", "intergenic_bases", "RnaSeqMetrics__INTRONIC_BASES", "intronic_bases", "RnaSeqMetrics__MEDIAN_3PRIME_BIAS", "median_3prime_bias", "RnaSeqMetrics__MEDIAN_5PRIME_BIAS", "median_5prime_bias", "RnaSeqMetrics__MEDIAN_5PRIME_TO_3PRIME_BIAS", "median_5prime_to_3prime_bias", "RnaSeqMetrics__MEDIAN_CV_COVERAGE", "median_cv_coverage", "RnaSeqMetrics__PCT_CODING_BASES", "pct_coding_bases", "RnaSeqMetrics__PCT_CORRECT_STRAND_READS", "pct_correct_strand_reads", "RnaSeqMetrics__PCT_INTERGENIC_BASES", "pct_intergenic_bases", "RnaSeqMetrics__PCT_INTRONIC_BASES", "pct_intronic_bases", "RnaSeqMetrics__PCT_MRNA_BASES", "pct_mrna_bases", "RnaSeqMetrics__PCT_RIBOSOMAL_BASES","pct_ribosomal_bases",  "RnaSeqMetrics__PCT_USABLE_BASES", "pct_usable_bases", "RnaSeqMetrics__PCT_UTR_BASES", "pct_utr_bases", "RnaSeqMetrics__PF_ALIGNED_BASES", "pct_pf_reads_aligned", "RnaSeqMetrics__PF_BASES", "pf_bases", "RnaSeqMetrics__RIBOSOMAL_BASES", "ribosomal_bases", "RnaSeqMetrics__UTR_BASES" , "utr_bases") 

Meta_Final <- as.data.frame( Meta_Final[, ORD ])

###Annotate with Ages Uncensord 
Master_Censor <- openxlsx::read.xlsx( synGet('syn18693152')$path, sheet=1)
Other_Censor <- openxlsx::read.xlsx( synGet('syn18632500')$path, sheet=1)
AgesUnCen <- read.csv( synGet('syn7116000')$path )

Comb_Uncensor <- as.data.frame( rbind( AgesUnCen[, c( 'projid', 'age_death' )] ,
                                       Other_Censor[, c( 'projid', 'age_death' ) ],
                                       Master_Censor[, c( 'projid','age_death' ) ] 
                                       )
                                )

Comb_Uncensor <- Comb_Uncensor[ as.character(Comb_Uncensor$projid) %in% as.character(Meta_Final$projid_repaired), ]
Comb_Uncensor <- Comb_Uncensor[!duplicated(Comb_Uncensor), ]
Comb_Uncensor$age_death <- signif( as.numeric( Comb_Uncensor$age_death ), digits = 10 )
Comb_Uncensor <- Comb_Uncensor[!duplicated(Comb_Uncensor), ]

row.names(Comb_Uncensor) <- Comb_Uncensor$projid

NotInAgesCensored <- NULL
NotInAgesCensored_90Plus <- NULL

for( i in 1:dim(Meta_Final) ){
  if( as.character( Meta_Final$projid )[ i ] %in% row.names(Comb_Uncensor) ){
    Meta_Final[ i, ]$age_death <- as.character( Comb_Uncensor[ as.character( Meta_Final$projid )[ i ], ]$age_death)
    #Comb_Uncensor[ as.character( Meta_Final$projid )[ i ], ]
  }else{
    if( as.character( Meta_Final$projid_repaired )[ i ] %in% row.names(Comb_Uncensor) ){
      Meta_Final[ i, ]$age_death <- as.character( Comb_Uncensor[ as.character( Meta_Final$projid_repaired )[ i ], ]$age_death)
    }else{
      NotInAgesCensored <- c( NotInAgesCensored, as.character( Meta_Final$projid )[ i ] )
      if( as.character( Meta_Final$age_death )[ i ] == "90+" ){
        NotInAgesCensored_90Plus <- c( NotInAgesCensored_90Plus, as.character( Meta_Final$projid )[ i ] )
      }
    }
  }
}

Passed_Master <- Master_Censor[ is.na(Master_Censor$age_death)==F, ]
table( Meta_Final$projid_repaired %in% Master_Censor$projid )
table( Master_Censor$projid %in% Meta_Final$projid_repaired )
Passed_Master <- Passed_Master[ Passed_Master$projid %in% Meta_Final$projid_repaired, ]



#Write Final MetaData to final file
write.csv( Meta_Final, 'RosMap_RNASeq_Meta_Master.csv', row.names = F )
activity <- Activity(
    'RosMap RNASeq Master With Ages Over 90',
    description='Total RNASeq Meta Data for RosMap RNASeq',
    executed = thisFile,
    used=c('syn22355545',"syn22314230","syn22301601","syn22283382","syn22283384","syn22301603","syn22314232","syn3191087",'syn21323366','syn21088596'))
file <- File('RosMap_RNASeq_Meta_Master.csv', description='RosMap RNASeq MetaData', executed = thisFile, parent='syn23569327')
file <- synStore(file, activity=activity)

file <- File('RosMap_RNASeq_Meta_Master.csv', description='RosMap RNASeq MetaData', executed = thisFile, parent='syn23573926')
file <- synStore(file, activity=activity)


####### Combine the counts across data cuts

#Get Count Data
Ros_DC1 <- get_data( 'syn22283382' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t()

Ros_DC2 <- get_data( 'syn22301601' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t() 

Ros_DC3 <- get_data( 'syn22314230' )[-c(1:4),] %>%
    tibble::column_to_rownames(var = "feature") %>%
    t() %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "sample") %>%
    tibble::column_to_rownames(var = "sample") %>%
    t()

   t()

total_counts <- cbind( Ros_DC1, Ros_DC2, Ros_DC3 )

total_counts <- cbind( feature=row.names(total_counts), total_counts)


write.table( total_counts, 'RosMap_totalcounts.txt', row.names = F, col.names = T, sep='\t', quote=F )
activity <- Activity(
    'RosMap Counts',
    description='All three Data cuts of the RosMap Data',
    executed = thisFile,
    used=c('syn22314230','syn22301601','syn22283382'))
file <- File('RosMap_totalcounts.txt', description='RosMap Counts', executed = thisFile, parent='syn23573926')
file <- synStore(file, activity=activity)

