---
title: "Run the drake plan"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Run the drake plan}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

1. [Write your own config file](https://kelshmo.github.io/sageseqr/articles/customize-config.html).

2. Specify the active configuration by setting `R_CONFIG_ACTIVE`.

```{r config-setup}
Sys.setenv(R_CONFIG_ACTIVE = "rosmap")
```

3. Load the `sageseqr` library and login to [Synapse](https://www.synapse.org/). `rnaseq_plan()` calls the arguments from the config file and creates the `drake` plan. Execute `drake::make(plan)` to compute. It is required to explicitly specify the name of the cache created with `drake::new_cache()` when multiple plans are executed in the same R session. Run this code from your project root.

```{r run-plan}
library(sageseqr)

# Login to Synapse. Make a Synapse account and use synapser to login: https://r-docs.synapse.org/articles/manageSynapseCredentials.html
synapser::synLogin()
setwd('~/RosMap_RNASeq/')


Sys.setenv(R_CONFIG_ACTIVE = "rosmap")


# Run the analysis
plan <- sageseqr::rnaseq_plan(
  metadata_id = config::get("metadata")$synID,
  metadata_version = config::get("metadata")$version,
  counts_id = config::get("counts")$synID,
  counts_version = config::get("counts")$version,
  gene_id_input = config::get("counts")$`gene id`,
  sample_id_input = config::get("metadata")$`sample id`,
  factor_input = config::get("factors"),
  continuous_input = config::get("continuous"),
  biomart_id = config::get("biomart")$synID,
  biomart_version = config::get("biomart")$version,
  filters = config::get("biomart")$filters,
  host = config::get("biomart")$host,
  organism = config::get("biomart")$organism, 
  conditions = config::get("conditions"), 
  x_var_for_plot = config::get("x_var"), 
  sex_var = config::get("sex check"), 
  color = config::get("dimensions")$color, 
  shape = config::get("dimensions")$shape,
  size = config::get("dimensions")$size,
  report_name = config::get("report"),
  skip_model = config::get("skip model")
)
```

```{r brokenoutplan}

MakeBioMart=FALSE

import_metadata <- get_data("syn23573928", 2L)
import_counts <- get_data("syn23593968", 1L)
counts <- tibble::column_to_rownames(import_counts, var = "feature")

clean_md <- clean_covariates(md = import_metadata, factors = c("sampleid", "apoe4", "batch", "notes", "tissue", "diagnosis", "sex", "race", "spanish", "braaksc", "ceradsc", "cogdx", "dcfdx_lv"), continuous = c("pmi", "rincontinuous", "rin2", "age_death", "cts_mmse30_lv", "pct_pf_reads_aligned", "pct_coding_bases", "pct_intergenic_bases", "pct_intronic_bases", "pct_ribosomal_bases"), sample_identifier = "sampleid")

if( MakeBioMart=TRUE ){
  biomart_results <- get_biomart(count_df = counts, synid = NULL, version = NULL, filters = "ensembl_gene_id", host = "ensembl.org", organism = "hsa")

  biomart_results$ensembl_gene_id <- row.names(biomart_results)
  biomart_results<- biomart_results[,c('ensembl_gene_id',	'hgnc_symbol',	
                                       'percentage_gene_gc_content', 'gene_biotype',
                                       'chromosome_name',	'gene_length')
                                    ]
  write.csv(biomart_results, 'RosMap_BioMartObject.csv', row.names=F, quote=F)
  
  thisFileName <- 'run-the-plan.Rmd'
  # Github link
  thisRepo <- githubr::getRepo(repository = "jgockley62/RosMap_RNASeq", ref="branch", refName='master')
  thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))


  parentID <- 'syn23573926'
  
  activity <- Activity(
    'RosMap Biomart Object',
    description='Gene ID (ensmbl) Info for RosMap RNA-Seq',
    executed = thisFile,
    used=c('syn23593968'))
file <- File('RosMap_BioMartObject.csv', description='Gene ID (ensmbl) Info for RosMap RNA-Seq', executed = thisFile, parent=parentID)
file <- synStore(file, activity=activity)

}
rm(count_df)
rm(synid)
rm(version)
rm(filters)
rm(host)
rm(organism)

function (count_df, synid, version, host, filters, organism) 
{
    if (is.null(synid)) {
        ensembl <- biomart_obj(organism, host)
        count_df <- parse_counts(count_df)
        gene_ids <- convert_geneids(count_df)
        message(paste0("Downloading sequence", ifelse(length(gene_ids) > 
            1, "s", ""), " ..."))
        if (length(gene_ids) > 100) 
            message("This may take a few minutes ...")
        attrs <- c(filters, "ensembl_exon_id", "chromosome_name", 
            "exon_chrom_start", "exon_chrom_end")
        coords <- biomaRt::getBM(filters = filters, attributes = attrs, 
            values = gene_ids, mart = ensembl, useCache = FALSE)
        gene_ids <- unique(coords[, filters])
        coords <- sapply(gene_ids, function(i) {
            i.coords <- coords[coords[, 1] == i, 3:5]
            g <- GenomicRanges::GRanges(i.coords[, 1], IRanges::IRanges(i.coords[, 
                2], i.coords[, 3]))
            g
        })
        length <- plyr::ldply(coords[gene_ids], function(x) sum(IRanges::width(x)), 
            .id = "ensembl_gene_id") %>% dplyr::rename(gene_length = .data$V1)
        gc_content <- biomaRt::getBM(filters = filters, attributes = c(filters, 
            "hgnc_symbol", "percentage_gene_gc_content", "gene_biotype", 
            "chromosome_name"), values = gene_ids, mart = ensembl, 
            useCache = FALSE)
        biomart_results <- dplyr::full_join(gc_content, length)
        biomart_results <- collapse_duplicate_hgnc_symbol(biomart_results)
        biomart_results <- tibble::column_to_rownames(biomart_results, 
            var = filters)
        biomart_results
    }
    else {
        biomart_results <- get_data(synid, version)
        biomart_results <- tibble::column_to_rownames(biomart_results, 
            var = filters)
        required_variables <- c("gene_length", "percentage_gene_gc_content")
        if (!all(required_variables %in% colnames(biomart_results))) {
            vars <- glue::glue_collapse(setdiff(required_variables, 
                colnames(biomart_results)), sep = ", ", last = " and ")
            message(glue::glue("Warning: {vars} missing from biomart object.\n                         This information is required for Conditional\n                         Quantile Normalization"))
        }
        return(biomart_results)
    }
}

```

```{r fulldrakeplan}
drake::make(
  plan,
  cache = drake::new_cache(
    config::get(
      "analysis title"
      )
    )
  )
```

4. Visualize the results of your work. 

```{r visualize}
drake::vis_drake_graph(
  plan,
  targets_only = TRUE
  )
```
